File: python-parser/utils/error_handler.py
"""
Standardized Error Handling for Python Parser Service
Provides consistent error responses and logging
"""

import logging
import traceback
from enum import Enum
from typing import Dict, Any, Optional, Tuple
from datetime import datetime
from flask import jsonify

class ErrorType(Enum):
    """Standard error types for categorization"""
    VALIDATION = "VALIDATION"
    NOT_FOUND = "NOT_FOUND"
    TIMEOUT = "TIMEOUT"
    PAYLOAD_TOO_LARGE = "PAYLOAD_TOO_LARGE"
    EXTERNAL_SERVICE = "EXTERNAL_SERVICE"
    INTERNAL = "INTERNAL"
    SECURITY = "SECURITY"
    RESOURCE_LIMIT = "RESOURCE_LIMIT"

class ErrorCode:
    """HTTP status codes for different error types"""
    ERROR_CODES = {
        ErrorType.VALIDATION: 400,
        ErrorType.NOT_FOUND: 404,
        ErrorType.TIMEOUT: 408,
        ErrorType.PAYLOAD_TOO_LARGE: 413,
        ErrorType.EXTERNAL_SERVICE: 502,
        ErrorType.INTERNAL: 500,
        ErrorType.SECURITY: 403,
        ErrorType.RESOURCE_LIMIT: 413
    }
    
    @classmethod
    def get_code(cls, error_type: ErrorType) -> int:
        return cls.ERROR_CODES.get(error_type, 500)

class AppError(Exception):
    """Application-specific error with standardized structure"""
    
    def __init__(self, message: str, error_type: ErrorType = ErrorType.INTERNAL, 
                 details: Optional[Dict[str, Any]] = None, cause: Optional[Exception] = None):
        super().__init__(message)
        self.message = message
        self.error_type = error_type
        self.status_code = ErrorCode.get_code(error_type)
        self.details = details or {}
        self.cause = cause
        self.timestamp = datetime.utcnow().isoformat() + 'Z'
        self.is_operational = True  # Distinguishes from programming errors

def create_error_response(error: Exception, request_id: Optional[str] = None) -> Tuple[Dict[str, Any], int]:
    """Creates standardized error response format"""
    
    if isinstance(error, AppError):
        app_error = error
    else:
        # Convert standard exceptions to AppError
        if isinstance(error, (FileNotFoundError, OSError)):
            app_error = AppError("Resource not found", ErrorType.NOT_FOUND, cause=error)
        elif isinstance(error, (MemoryError,)):
            app_error = AppError("Insufficient memory", ErrorType.RESOURCE_LIMIT, cause=error)
        elif isinstance(error, (TimeoutError,)):
            app_error = AppError("Operation timed out", ErrorType.TIMEOUT, cause=error)
        elif isinstance(error, (ValueError, TypeError)):
            app_error = AppError("Invalid input data", ErrorType.VALIDATION, cause=error)
        else:
            app_error = AppError("Internal server error", ErrorType.INTERNAL, cause=error)
    
    response = {
        "success": False,
        "error": {
            "message": app_error.message,
            "type": app_error.error_type.value,
            "timestamp": app_error.timestamp
        }
    }
    
    # Add details if available and not in production
    if app_error.details and not _is_production():
        response["error"]["details"] = app_error.details
        
    if request_id:
        response["error"]["requestId"] = request_id
    
    # Don't expose internal details in production
    if _is_production() and app_error.error_type == ErrorType.INTERNAL:
        response["error"]["message"] = "Internal server error"
    
    return response, app_error.status_code

def log_error(error: Exception, context: Optional[Dict[str, Any]] = None):
    """Logs errors with consistent format and appropriate level"""
    
    if isinstance(error, AppError):
        log_level = logging.ERROR if error.status_code >= 500 else logging.WARNING
        
        log_data = {
            "error_type": error.error_type.value,
            "status_code": error.status_code,
            "message": error.message,
            "timestamp": error.timestamp
        }
        
        if error.details:
            log_data["details"] = error.details
            
        if error.cause:
            log_data["cause"] = str(error.cause)
            
        if context:
            log_data["context"] = context
            
        logging.log(log_level, f"Application Error: {error.message}", extra=log_data)
        
        # Log stack trace for internal errors
        if error.error_type == ErrorType.INTERNAL:
            logging.error(f"Stack trace: {traceback.format_exc()}")
    else:
        logging.error(f"Unhandled exception: {str(error)}", exc_info=True)

def handle_error(func):
    """Decorator to handle errors in Flask route handlers"""
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except Exception as e:
            log_error(e, {"function": func.__name__})
            response, status_code = create_error_response(e)
            return jsonify(response), status_code
    
    wrapper.__name__ = func.__name__
    return wrapper

def _is_production() -> bool:
    """Check if running in production environment"""
    import os
    return os.getenv('FLASK_ENV') == 'production' or os.getenv('ENVIRONMENT') == 'production'

# Common error creators for frequent scenarios
def create_validation_error(message: str, field: Optional[str] = None) -> AppError:
    """Create validation error with optional field information"""
    details = {"field": field} if field else None
    return AppError(message, ErrorType.VALIDATION, details)

def create_security_error(message: str, threat_type: Optional[str] = None) -> AppError:
    """Create security-related error"""
    details = {"threat_type": threat_type} if threat_type else None
    return AppError(message, ErrorType.SECURITY, details)

def create_timeout_error(operation: str, timeout_seconds: Optional[int] = None) -> AppError:
    """Create timeout error with operation context"""
    details = {"timeout_seconds": timeout_seconds} if timeout_seconds else None
    return AppError(f"{operation} timed out", ErrorType.TIMEOUT, details)

def create_resource_limit_error(resource: str, limit: Optional[str] = None) -> AppError:
    """Create resource limit error"""
    details = {"limit": limit} if limit else None
    return AppError(f"{resource} limit exceeded", ErrorType.RESOURCE_LIMIT, details)

def create_external_service_error(service: str, cause: Optional[Exception] = None) -> AppError:
    """Create external service error"""
    return AppError(f"{service} service unavailable", ErrorType.EXTERNAL_SERVICE, cause=cause)

File: backend/utils/errorHandler.js
/**
 * Standardized Error Handling Utility
 * Provides consistent error responses across the backend service
 */

/**
 * Standard error types for better categorization
 */
export const ErrorTypes = {
  VALIDATION: 'VALIDATION',
  AUTHENTICATION: 'AUTHENTICATION', 
  AUTHORIZATION: 'AUTHORIZATION',
  NOT_FOUND: 'NOT_FOUND',
  RATE_LIMIT: 'RATE_LIMIT',
  EXTERNAL_SERVICE: 'EXTERNAL_SERVICE',
  INTERNAL: 'INTERNAL',
  TIMEOUT: 'TIMEOUT',
  PAYLOAD_TOO_LARGE: 'PAYLOAD_TOO_LARGE'
};

/**
 * Standard error codes mapping to HTTP status codes
 */
const ErrorCodes = {
  [ErrorTypes.VALIDATION]: 400,
  [ErrorTypes.AUTHENTICATION]: 401,
  [ErrorTypes.AUTHORIZATION]: 403,
  [ErrorTypes.NOT_FOUND]: 404,
  [ErrorTypes.RATE_LIMIT]: 429,
  [ErrorTypes.EXTERNAL_SERVICE]: 502,
  [ErrorTypes.INTERNAL]: 500,
  [ErrorTypes.TIMEOUT]: 408,
  [ErrorTypes.PAYLOAD_TOO_LARGE]: 413
};

/**
 * Application Error class for consistent error handling
 */
export class AppError extends Error {
  constructor(message, type = ErrorTypes.INTERNAL, details = null, cause = null) {
    super(message);
    this.name = 'AppError';
    this.type = type;
    this.statusCode = ErrorCodes[type] || 500;
    this.details = details;
    this.cause = cause;
    this.timestamp = new Date().toISOString();
    this.isOperational = true; // Marks this as an expected error vs programming error
    
    // Capture stack trace
    Error.captureStackTrace(this, AppError);
  }
}

/**
 * Creates standardized error response format
 */
export const createErrorResponse = (error, requestId = null) => {
  const response = {
    success: false,
    error: {
      message: error.message || 'An unexpected error occurred',
      type: error.type || ErrorTypes.INTERNAL,
      timestamp: error.timestamp || new Date().toISOString(),
      ...(error.details && { details: error.details }),
      ...(requestId && { requestId })
    }
  };
  
  // Don't expose internal error details in production
  if (process.env.NODE_ENV === 'production' && error.type === ErrorTypes.INTERNAL) {
    response.error.message = 'Internal server error';
    delete response.error.details;
  }
  
  return response;
};

/**
 * Express error handling middleware
 */
export const errorHandler = (logger) => (err, req, res, next) => {
  let error = err;
  
  // Convert non-AppError errors to AppError
  if (!(err instanceof AppError)) {
    // Handle specific known error types
    if (err.code === 'ENOTFOUND' || err.code === 'ECONNREFUSED') {
      error = new AppError('External service unavailable', ErrorTypes.EXTERNAL_SERVICE, null, err);
    } else if (err.code === 'ETIMEDOUT' || err.type === 'timeout') {
      error = new AppError('Request timeout', ErrorTypes.TIMEOUT, null, err);
    } else if (err.type === 'entity.too.large') {
      error = new AppError('Request payload too large', ErrorTypes.PAYLOAD_TOO_LARGE, null, err);
    } else if (err.type === 'entity.parse.failed' || err instanceof SyntaxError || /Unexpected token/.test(err.message || '')) {
      // JSON parse errors from body-parser should be treated as validation failures
      error = new AppError('Invalid JSON payload', ErrorTypes.VALIDATION, null, err);
    } else {
      error = new AppError(err.message || 'Internal server error', ErrorTypes.INTERNAL, null, err);
    }
  }
  
  // Log error details
  const logLevel = error.statusCode >= 500 ? 'error' : 'warn';
  logger[logLevel](`${error.type} (${error.statusCode}): ${error.message}`, {
    error: {
      name: error.name,
      type: error.type,
      statusCode: error.statusCode,
      stack: error.stack,
      cause: error.cause?.message,
      details: error.details
    },
    request: {
      method: req.method,
      url: req.url,
      ip: req.ip,
      userAgent: req.get('User-Agent')
    }
  });
  
  // Send standardized error response
  const requestId = req.id || req.headers['x-request-id'];
  const errorResponse = createErrorResponse(error, requestId);
  
  // For validation errors (tests expect a simple string in .body.error), provide compatibility
  if (error.type === ErrorTypes.VALIDATION || error.type === ErrorTypes.PAYLOAD_TOO_LARGE) {
    // Some legacy endpoints (file uploads under /analyze) expect a success:false envelope
    try {
      const p = req.path || req.originalUrl || '';
      if (typeof p === 'string' && p.startsWith('/analyze')) {
        return res.status(error.statusCode).json({ success: false, error: error.message });
      }
    } catch (e) {
      // ignore and fallthrough
    }
    // Default: return the simple string error for other endpoints (tests expect this)
    return res.status(error.statusCode).json({ error: error.message });
  }

  res.status(error.statusCode).json(errorResponse);
};

/**
 * Async wrapper to catch errors in async route handlers
 */
export const asyncHandler = (fn) => (req, res, next) => {
  Promise.resolve(fn(req, res, next)).catch(next);
};

/**
 * Common error creators for frequent scenarios
 */
export const createValidationError = (message, details = null) => 
  new AppError(message, ErrorTypes.VALIDATION, details);

export const createNotFoundError = (resource = 'Resource') => 
  new AppError(`${resource} not found`, ErrorTypes.NOT_FOUND);

export const createTimeoutError = (operation = 'Operation') => 
  new AppError(`${operation} timed out`, ErrorTypes.TIMEOUT);

export const createExternalServiceError = (service, originalError = null) => 
  new AppError(`${service} service unavailable`, ErrorTypes.EXTERNAL_SERVICE, null, originalError);